Source/JavaScriptCore/wasm/WasmStruct.cpp

```diff
@@ -0,0 +1,112 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WasmStruct.h"
+
+#if ENABLE(WEBASSEMBLY)
+
+namespace JSC::Wasm {
+
+constexpr size_t typeKindSize(TypeKind kind)
+{
+    switch (kind) {
+    case TypeKind::I32:
+    case TypeKind::F32: {
+        return 32;
+    }
+    case TypeKind::I64:
+    case TypeKind::F64:
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull: {
+        return 64;
+    }
+    case TypeKind::Func:
+    case TypeKind::Struct:
+    case TypeKind::Void:
+    case TypeKind::I31ref: {
+        break;
+    }
+    }
+
+    ASSERT_NOT_REACHED();
+    return 0;
+}
+
+size_t WasmStruct::structDataAllocationSize(const StructType* structType)
+{
+    size_t totalSize = 0u;
+    for (StructFieldCount i = 0; i < structType->fieldCount(); ++i)
+        totalSize += typeKindSize(structType->field(i).type.kind);
+    return totalSize;
+}
+
+WasmStruct::WasmStruct(Ref<TypeDefinition>&& type, FixedVector<uint64_t>&& arguments)
+    : m_type(WTFMove(type))
+{
+    for (unsigned fieldIndex = 0; fieldIndex < structType().fieldCount(); ++fieldIndex)
+        set(fieldIndex, arguments[fieldIndex]);
+}
+
+uint64_t WasmStruct::get(uint32_t fieldIndex) const
+{
+    ASSERT(fieldIndex < structType().fieldCount());
+    const uint8_t* targetPointer = reinterpret_cast<const uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex);
+    uint64_t payload = 0;
+    if (typeKindSize(fieldKind(fieldIndex)) == 64)
+        payload = *reinterpret_cast<const uint64_t*>(targetPointer);
+    else {
+        ASSERT(typeKindSize(fieldKind(fieldIndex)) == 32);
+        payload = *reinterpret_cast<const uint32_t*>(targetPointer);
+    }
+    return payload;
+}
+
+size_t WasmStruct::getFieldOffset(uint32_t fieldIndex) const
+{
+    ASSERT(fieldIndex < structType().fieldCount());
+
+    size_t fieldOffset = 0;
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSize(fieldKind(i));
+    return fieldOffset;
+}
+
+void WasmStruct::set(uint32_t fieldIndex, uint64_t value)
+{
+    uint8_t* targetPointer = reinterpret_cast<uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex);
```

Comment:
What is the plan for WASM gc write barriers? If a field with a reference type will just store the JSValue, don't we need a write barrier here?

Reply:
added for reference types